source/server-discovery-and-monitoring/tests/README.rst

```diff
@@ -262,6 +262,32 @@ Run the following test(s) on MongoDB 4.4+.
              mode: "off",
          });
 
+Heartbeat Tests
+~~~~~~~~~~~~~~~
+
+1. Test that ``ServerHeartbeatStartedEvent`` is emitted after the monitoring socket was created and before the ``hello`` is sent to the server
+
+   #. Create a mock TCP server (example shown below) that pushes a ``client connected`` event to a shared array when a client connects and a ``client hello received`` event when the server receives the client hello and then closes the connection::
+         
+        let events = [];
+        server = createServer(clientSocket => {
+          events.push('client connected');
+
+          clientSocket.on('data', () => {
+            events.push('client hello received');
+            clientSocket.end();
+          });
+        });
+        server.listen(9999);
+
+   #. Create a client with ``serverSelectionTimeoutMS: 500`` and listen to ``ServerHeartbeatStartedEvent`` and ``ServerHeartbeatFailedEvent``, pushing the event name to the same shared array as the mock TCP server.
+
+   #. Attempt to connect client to previously created TCP server, catching the error when the client fails to connect 
+
+   #. Assert that the event array has the following contents in the same order:: 
+
+        ['serverHeartbeatStartedEvent', 'client connected', 'client hello received', 'serverHeartbeatFailedEvent']
```

Comment:
@ShaneHarvey 

Since we're concerned with the the fact that the `ServerHeartbeatStartedEvent` gets emitted before the connection is created and so we should fail if the connection fails to be established, would it be sufficient for us to just try to connect to a port that has nothing listening on it and expect to get a `ServerHeartbeatStartedEvent` followed by a `ServerHeartbeatFailedEvent`, or do we want to keep this granularity?

Reply:
Yes I think that test would work as well. Would it be easier to implement?

Reply:
Yes, it would. It wouldn't require the mock server at all.

Reply:
Revisiting this test; unless a driver can spy on the function that creates the socket, the updated version of this test wouldn't prevent something like this from passing:

```js
let socketOrError = await connect().catch(error => error);

if (isSocket(socketOrError)) {
  emitHBStarted()
  await runHeartBeat()
} else {
  emitHBStarted()
  throw socketOrError;
}
// Assume emitHBFailed will be done if any of the above throws
```

We should roll back to actually using the mock server as that will provide a stronger guarantee on the ordering we want to test here.